### PR TITLE
feat(function-v2): Add calc domain for comparison.

### LINKED ISSUE #8739: Feature: Add calc_domain for comparison functions
**Summary**

Currently, `eq`, `gt`, and other comparison functions in the new expression framework have no `calc_domain` definitions. However, they are important in datablock pruning, and thus we're going to add them back.

### PR DESCRIPTION
**Pull request by RinChanNOWWW**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

Closes #8739

### CODE REVIEW

**Review by TCeason — APPROVED** [review]
LGTM

File: src/query/functions-v2/src/scalars/comparison.rs [hunk 1]
```diff
@@ -39,89 +39,259 @@ use regex::bytes::Regex;
...
+        has_true: false,
+        has_false: true,
+    }
+}
+
+fn register_string_cmp(registry: &mut FunctionRegistry) {
+    registry.register_2_arg::<StringType, StringType, BooleanType, _, _>(
         "eq",
         FunctionProperty::default(),
         |_, _| FunctionDomain::NoThrow,
```
Comment (sundy-li) [inline src/query/functions-v2/src/scalars/comparison.rs:70]: @andylokandy  Is it recommended to use `FunctionDomain::Domain(BooleanDomain {
                has_false: true,
                has_true: true,
            })`  or  `FunctionDomain::NoThrow` ?

**Comment by Xuanwo** [comment]
@mergify update

File: src/query/functions-v2/src/scalars/comparison.rs (hunk 1 above)
Comment (andylokandy) [reply src/query/functions-v2/src/scalars/comparison.rs:70]: `FunctionDomain::NoThrow` is equivalent to `FunctionDomain::Domain(T::full_domain())`.

File: src/query/functions-v2/src/scalars/comparison.rs [hunk 2]
```diff
@@ -39,89 +39,259 @@ use regex::bytes::Regex;
...
+                    FunctionDomain::Domain(all_false_domain())
+                } else if d1.min == d1.max && d2.min == d2.max {
+                    if d1.min == d2.min {
+                        FunctionDomain::Domain(all_true_domain())
+                    } else {
+                        FunctionDomain::Domain(all_false_domain())
+                    }
+                } else {
+                    FunctionDomain::NoThrow
+                }
```
Comment (andylokandy) [inline src/query/functions-v2/src/scalars/comparison.rs:121]:
```suggestion
                if d1.min > d2.max || d1.max < d2.min {
                    FunctionDomain::Domain(all_false_domain())
                } else if d1.min == d1.max && d2.min == d2.max && d1.min == d2.min {
                        FunctionDomain::Domain(all_true_domain())
                } else {
                    FunctionDomain::NoThrow
                }
```

Comment (andylokandy) [reply src/query/functions-v2/src/scalars/comparison.rs:121]: L117 is covered by L112

Comment (andylokandy) [inline src/query/functions-v2/src/scalars/comparison.rs:121]: therefore, the following is sufficient:
```suggestion
                if d1.min > d2.max || d1.max < d2.min {
                    FunctionDomain::Domain(all_false_domain())
                } else {
                    FunctionDomain::NoThrow
                }
```

**Review by andylokandy — APPROVED** [review]

File: src/query/functions-v2/src/scalars/comparison.rs (hunk 1 above)
Comment (sundy-li) [reply src/query/functions-v2/src/scalars/comparison.rs:64]: Yes, let's do this in another pr. Currently it's ok in this one.

**Review by sundy-li — APPROVED** [review]

**Pull request merged**